zcash_primitives/src/note_encryption.rs

```diff
@@ -405,7 +439,15 @@ pub fn try_sapling_note_decryption(
         NOTE_PLAINTEXT_SIZE
     );
 
-    let (note, to) = parse_note_plaintext_without_memo(ivk, cmu, &plaintext)?;
+    let (note, to) = parse_note_plaintext_without_memo::<P>(height, ivk, cmu, &plaintext)?;
+
+    match note.rseed {
+        Rseed::AfterZip212(rseed) => {
+            let derived_esk = Fs::to_uniform(prf_expand(&rseed, &[0x05]).as_bytes());
+            assert_eq!(note.g_d.mul(derived_esk, &JUBJUB), *epk);
```

Comment:
To match the [consensus-critical implementation](https://github.com/zcash/zcash/blob/67ab09ae43caf17c0331b5c7a0aa26e08ee252b2/src/zcash/Note.cpp#L294-L305) in `zcashd`:
```suggestion
            if note.g_d.mul(derived_esk, &JUBJUB) !=epk {
                return None;
            }
```
(`zcashd` compares the encodings of `[esk] g_d` and `epk`, but both `note.g_d` and `epk` are constrained by the type system to be in the prime-order subgroup, so comparing the points is equivalent to comparing their encodings.)

Reply:
There might be a discrepancy with the spec here. The type of `epk` in the protocol spec is `KA.Public`, not `KA.PublicPrimeOrder`, but the actual code pathways used in `zcashd` from ZIP 212 onward enforce it to be `KA.PublicPrimeOrder`:
- `Diversifier::g_d` returns `edwards::Point<_, PrimeOrder>`
- `g_d.mul(esk)` therefore also returns `edwards::Point<_, PrimeOrder>` ([`edwards::Point::mul` returns `Self`](https://github.com/zcash/librustzcash/blob/139fc09f1028f0f2d073748dfca8529dbe98807b/zcash_primitives/src/jubjub/edwards.rs#L465)).
- `zcashd` compares `epk` against the output of `librustzcash_sapling_ka_derivepublic`, which internally encodes the result of `g_d.mul(esk)`.

Reply:
if KA<sup>Sapling</sup>.DerivePublic(esk, g<sub>d</sub>) indeed returns a point in the prime-order subgroup. This is not a bug in the signature of DerivePublic; it holds only because g<sub>d</sub> is recomputed as an output of GroupHash<sup>J<sup>(r)*</sup></sup>.

There is also nothing preventing esk from being zero. (The sender must know esk due to the "Ephemeral public key integrity" check in the Output circuit, but that does not prevent esk from being zero.)

It's not correct to say that comparing the points is equivalent to comparing their encodings. The prime-order subgroup includes the zero point, and the zero point has two valid encodings under abst<sup>J</sup> (with different compressed sign bits). Since `epk` is given in the Output description, it could have a non-canonical encoding. Whether this matters depends on the implementation of the Output circuit "Ephemeral public key integrity" check.

Reply:
When epk is represented as an input to the Output circuit, it is taken as two field elements u and v, i.e. in uncompressed affine representation. So the Output circuit cannot "see" that epk was originally represented non-canonically. It follows that the current implementation of Canopy in zcashd that is active on testnet is nonconformant to the spec (both as written and as intended), since it will accept a non-canonical representation of the zero point as epk if esk = 0. We need to decide whether to change the spec or the implementation.